### ISSUE TITLE
[REQUEST] i want to know how to merge  deepspeed multi gpu optim file into one pytorch optim.pt file ?  

### ISSUE THREAD

**Issue by ldh127** [microsoft/DeepSpeed#5460]

**Comment by loadams** [comment]
Hi @ldh127 - can you please be more specific, share more about what you are trying to do and what errors you are hitting?

**Comment by ldh127** [comment]
yes，but i think this code is for ds2universe model param ,not for merging multi optim file into to one file , it can&nbsp; process&nbsp;merge deepspeed multi gpu optim file into one pytorch optim.pt file?&nbsp;


ldh
***@***.***



&nbsp;




------------------&nbsp;原始邮件&nbsp;------------------
发件人: ***@***.***&gt;; 
发送时间: 2024年4月25日(星期四) 凌晨3:40
收件人: ***@***.***&gt;; 
抄送: ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [microsoft/DeepSpeed] [REQUEST] i want to know how to merge  deepspeed multi gpu optim file into one pytorch optim.pt file ?   (Issue #5460)














 
@ldh127, does the following help?
 https://deepspeed.readthedocs.io/en/latest/model-checkpointing.html#zero-checkpoint-fp32-weights-recovery

**Comment by ldh127** [comment]
> Hi @ldh127 - can you please be more specific, share more about what you are trying to do and what errors you are hitting?

yes ,i use transformers trainer to call deepspeed , it save the deepspeed checkpoint which contains  multi gpu model and optim file , i want just one file optim.pt file to choosing  sft data , my code can only load one global optim.pt , but deepspeed checkpoint get multi part optim and model file , how can i. merge multi optim file to one global file ?

**Comment by tjruwase** [comment]
> yes，but i think this code is for ds2universe model param ,not for merging multi optim file into to one file , it can&nbsp; process&nbsp;merge deepspeed multi gpu optim file into one pytorch optim.pt file?&nbsp; ldh ***@***.*** &nbsp;
> […](#)

@ldh127, why do you say the link is related to ds2universal? Did you try it? Can you clarify how your scenario is different from the use case below? Thanks!

<img width="534" alt="image" src="https://github.com/microsoft/DeepSpeed/assets/4271600/96a97e47-6c27-4693-9875-de9d89117a4a">

**Comment by ldh127** [comment]
> > yes，but i think this code is for ds2universe model param ,not for merging multi optim file into to one file , it can  process merge deepspeed multi gpu optim file into one pytorch optim.pt file?  ldh _**@**_.***  
> > […](#)
> 
> @ldh127, why do you say the link is related to ds2universal? Did you try it? Can you clarify how your scenario is different from the use case below? Thanks!
> 
> <img alt="image" width="534" src="https://private-user-images.githubusercontent.com/4271600/326058623-96a97e47-6c27-4693-9875-de9d89117a4a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQyMTk0NjksIm5iZiI6MTcxNDIxOTE2OSwicGF0aCI6Ii80MjcxNjAwLzMyNjA1ODYyMy05NmE5N2U0Ny02YzI3LTQ2OTMtOTg3NS1kZTlkODkxMTdhNGEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDQyNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA0MjdUMTE1OTI5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9N2QxM2E2ZmZmMTJjYzE4YWZhNmQ1MDMxNWIxOGJmZjVjN2FjODJjZDFlMmI2ODcyMGIyNDExYzAzZGY4ZDYxYSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.3zBaokmJILyQRtAJr9wv3l2dz2nuFJq60UOWhhCqBTk">

yes , i try this code ， finally i surely get only one .pth file, but you can see my details,  i 
![0ce59e0b-7dfd-4743-987f-608d008b4a66](https://github.com/microsoft/DeepSpeed/assets/12208944/1449e0d3-57da-481f-bf5b-b62f8da05a4e)
this is my deepspeed checkpoint file, i use your code to read this folder ,and finally it merge and save only one file , i use this code ,you can see 
![807e76d9-504f-413e-bff6-f7ba86a2f28e](https://github.com/microsoft/DeepSpeed/assets/12208944/e3da3773-ccfa-4562-bcea-4a6cf91116f0)
,and i get the file like this , 
![01e94eb8-fe5f-4737-a2fb-aef60c8c9966](https://github.com/microsoft/DeepSpeed/assets/12208944/2f884a39-37e4-47c3-835f-31a1806efaa6)
, you can see that i print the state_dict name, it is like base_model.model.model.layers.38.self_attn.q_proj.lora_A.default.weight
base_model.model.model.layers.38.self_attn.q_proj.lora_B.default.weight
base_model.model.model.layers.38.self_attn.k_proj.lora_A.default.weight
base_model.model.model.layers.38.self_attn.k_proj.lora_B.default.weight
, but it seems the models name ,not the optim file name ?

**Comment by ldh127** [comment]
> > yes，but i think this code is for ds2universe model param ,not for merging multi optim file into to one file , it can  process merge deepspeed multi gpu optim file into one pytorch optim.pt file?  ldh _**@**_.***  
> > […](#)
> 
> @ldh127, why do you say the link is related to ds2universal? Did you try it? Can you clarify how your scenario is different from the use case below? Thanks!
> 
> <img alt="image" width="534" src="https://private-user-images.githubusercontent.com/4271600/326058623-96a97e47-6c27-4693-9875-de9d89117a4a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQyMTk0NjksIm5iZiI6MTcxNDIxOTE2OSwicGF0aCI6Ii80MjcxNjAwLzMyNjA1ODYyMy05NmE5N2U0Ny02YzI3LTQ2OTMtOTg3NS1kZTlkODkxMTdhNGEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDQyNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA0MjdUMTE1OTI5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9N2QxM2E2ZmZmMTJjYzE4YWZhNmQ1MDMxNWIxOGJmZjVjN2FjODJjZDFlMmI2ODcyMGIyNDExYzAzZGY4ZDYxYSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.3zBaokmJILyQRtAJr9wv3l2dz2nuFJq60UOWhhCqBTk">

you can see the picture , if the finally file which named demo_state_dict.pth contains optim param ,but how can i get the optim state_dict ? if it is the merged optim file , it seems i can use state_dict["optim_state"] like this way to get the dict ,but it has no optim_state key in the dict , so  i donot konw what error in my operate steps

**Comment by ldh127** [comment]
> > yes，but i think this code is for ds2universe model param ,not for merging multi optim file into to one file , it can  process merge deepspeed multi gpu optim file into one pytorch optim.pt file?  ldh _**@**_.***  
> > […](#)
> 
> @ldh127, why do you say the link is related to ds2universal? Did you try it? Can you clarify how your scenario is different from the use case below? Thanks!
> 
> <img alt="image" width="534" src="https://private-user-images.githubusercontent.com/4271600/326058623-96a97e47-6c27-4693-9875-de9d89117a4a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQyMjE0NjAsIm5iZiI6MTcxNDIyMTE2MCwicGF0aCI6Ii80MjcxNjAwLzMyNjA1ODYyMy05NmE5N2U0Ny02YzI3LTQ2OTMtOTg3NS1kZTlkODkxMTdhNGEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDQyNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA0MjdUMTIzMjQwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZmU5MWFiNmI3ZWI4MjEwOTI0MjE5NGZkNjRjNGE5ZDVmOWZhNTIyZTgxMTI1OWRmZjY3MWMwY2JiMmE4ZWUyMCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.patF9Z2G4Tc_rkFxFPIua517s2cBPfpFOKVMAs8ptSs">

i also read the code in this url: https://github.com/microsoft/DeepSpeed/blob/4c15ad9f8d51a1950842c69bbbc9d93c73afbcfc/deepspeed/utils/zero_to_fp32.py , but i do not know if i need to update what code , can you give me more detail help? thanks , need some ditail
